Core/clim-core/panes/composition.lisp

```diff
@@ -1152,72 +1161,33 @@
                                         ; if this is the right place, reusing the above calculation might be a good idea
-      (allocate-space child
-                      (max child-min-width child-width  width)
-                      (max child-min-height child-height height)))))
+    (allocate-space child
+                    (max child-min-width child-width  width)
+                    (max child-min-height child-height height))
+    (scroller-pane/update-scroll-bars parent)))
 
 (defmethod note-input-focus-changed ((pane viewport-pane) state)
   (note-input-focus-changed (sheet-child pane) state))
 
-;;; This method ensures that when the child changes size, the viewport
-;;; will move its focus so that it will not display a region outside
-;;; of `child' (if at all possible, this ideal can be circumvented by
-;;; creating a child sheet that is smaller than the viewport). I do
-;;; not believe having a viewport look at "empty" space is ever
-;;; useful.
 (defmethod note-space-requirements-changed ((pane viewport-pane) child)
-  (let ((viewport-width (bounding-rectangle-width pane))
-        (viewport-height (bounding-rectangle-height pane))
-        (child-width (bounding-rectangle-width child))
-        (child-height (bounding-rectangle-height child)))
-    (destructuring-bind (horizontal-scroll vertical-scroll)
-        (mapcar #'- (multiple-value-list
-                     (transform-position (sheet-transformation child) 0 0)))
-      ;; XXX: We cannot use `scroll-extent', because McCLIM ignores it
-      ;; unless the scrollee happens to be drawing. Very weird, should
-      ;; be fixed.
-
-      ;; It's not a bug, it's a feature. This requires further thought. -Hefner
-      (move-sheet child
-                  (- (if (> (+ horizontal-scroll viewport-width)
-                            child-width)
-                         (- child-width viewport-width)
-                         horizontal-scroll))
-                  (- (if (> (+ vertical-scroll viewport-height)
-                            child-height)
-                         (- child-height viewport-height)
-                         vertical-scroll)))
-      (scroller-pane/update-scroll-bars (sheet-parent pane)))))
-
-;;; Accounting for changed space requirements
-
-(defmethod change-space-requirements ((pane viewport-pane) &rest options)
-  (declare (ignore options))
-  (let* ((client (sheet-child pane))
-         (sr (compose-space client))
-         (width  (max (bounding-rectangle-width pane)
-                      (space-requirement-width sr)))
-         (height (max (bounding-rectangle-height pane)
-                      (space-requirement-height sr))))
-    (resize-sheet client width height)
-    (allocate-space client width height)
-    (scroller-pane/update-scroll-bars (sheet-parent pane))))
-
+  (multiple-value-bind (width height) (bounding-rectangle-size pane)
+    (allocate-space pane width height)))
```

Comment:
allocate-space is part of the layout-protocol, so if this function is invoked in the context of `changing-space-requirements`, this operation should be batched instead of being eagerly invoked.

Reply:
but this is not a regression, I suppose it could be addressed when changing-space-requirements is improved